### ISSUE TITLE
Compilation error: expected primary-expression before token

### ISSUE THREAD

**Issue by GuillaumeDesforges** [rusty1s/pytorch_sparse#44]
I'm packaging pytorch_sparse for NixOS.

## Bug description
Version: revision 2eff407270c271424ecaf6b8fbcaa7dc1564ffac (current master HEAD atm)
When building I get many errors such as this one:
```
/build/pytorch_sparse-2eff407/csrc/cpu/spspmm_cpu.cpp:90:62: error: expected primary-expression before ‘>’ token
   90 |         valC_data = optional_valueC.value().data_ptr<scalar_t>();
      |                                                              ^
```

## Investigations

Could be an issue in my build setup but I don't have much knowledge about this issue with gcc.

Seems related to this:
* https://stackoverflow.com/questions/5802077/strange-gcc-error-expected-primary-expression-before-token
* https://stackoverflow.com/questions/610245/where-and-why-do-i-have-to-put-the-template-and-typename-keywords

## More

For the record:

```
torch-sparse-cuda = super.buildPythonPackage rec {
        pname = "torch_sparse";
        version = "0.6.0+cuda";

        doCheck = false;

        src = pkgs.fetchgit {
          url = "https://github.com/rusty1s/pytorch_sparse";
          rev = "2eff407270c271424ecaf6b8fbcaa7dc1564ffac";
          sha256 = "1cpjwkxlah0xrv9gv4m3bmm837zv3x5lxqdp4b90nb0am4zf8dsh";
        };

        nativeBuildInputs = with pkgs; [
          which
          gcc
        ];

        buildInputs = with super; [
          pytestrunner
        ];

        propagatedBuildInputs = with super; [
          pytorch
        ];
      };
```

**Comment by rusty1s** [comment]
Are you sure there are no other errors mentioned before this? I feel that I can not much do about this, this seems PyTorch/GCC related. Which version of GCC are you using?

**Comment by GuillaumeDesforges** [comment]
GCC version is 9.2.0
I'm digging the issue, brb with more info

**Comment by GuillaumeDesforges** [comment]
Here is the full log : https://pastebin.com/z4RxEYrm

**Comment by SteveJunGao** [comment]
do you have any ideas how to resolve this issue, I also encountered this problem

**Comment by matherm** [comment]
I just had the same problem and resolved it by upgrading pytorch to 1.5.0.

`conda install -c pytorch pytorch
export CUDA=cu92 # cuda 9.2. in my case
pip install torch-scatter==latest+${CUDA} torch-sparse==latest+${CUDA} -f https://pytorch-geometric.com/whl/torch-1.5.0.html
`

**Comment by GuillaumeDesforges** [comment]
Now that nixpkgs has update to pytorch 1.5.0, I should try again then
